File: Implementation/C#/Source/Text/Writer/IndentPrinter.cs
using TW = System.IO.TextWriter;

namespace Cks.Text.Writer {

/**
 * Printer that manages indentation.
 */
public class IndentPrinter : TW
{
	private readonly TW Out;
	private readonly string PrefixString;
	private readonly string IndentString;
	private uint Level = 0;

	private bool nl;

	private static readonly string DefaultIndent = "  ";

	public IndentPrinter(TW Out, string IndentString, string PrefixString, bool FreshLine)
	{
		this.Out = Out;
		this.IndentString = IndentString;
		this.PrefixString = PrefixString;
		this.nl = FreshLine;
	}

	public IndentPrinter(TW Out, string IndentString, string PrefixString)
		: this(Out, IndentString, PrefixString, true) {}

	public IndentPrinter(TW Out, string IndentString)
		: this(Out, IndentString, "") {}

	public IndentPrinter(TW Out)
		: this(Out, DefaultIndent) {}

	public override System.Text.Encoding Encoding { get { return Out.Encoding; } }

	// -----------------------------------------------
	// Indentation Control
	
	public void Indent(uint NumIndents)
	{
		this.Level += NumIndents;
	}

	public void Indent()
	{
		this.Level += 1;
	}

	public void Dedent(uint NumIndents)
	{
		if (NumIndents > this.Level) throw new System.ArgumentException("attempt to decrase indent by " + NumIndents + ", but indent level is currently " + Level);
		this.Level -= NumIndents;
	}

	public void Dedent()
	{
		if (this.Level < 1) throw new System.ArgumentException("attempt to decrase indent, but indent level is currently " + Level);
		this.Level -= 1;
	}

	// -----------------------------------------------
	// Print Functions
	
	public override void Write(bool v) { Pre(); Out.Write(v); }
	public override void Write(char v) { Pre(); Out.Write(v); }
	public override void Write(char[] v) { Pre(); Out.Write(v); }
	public override void Write(decimal v) { Pre(); Out.Write(v); }
	public override void Write(double v) { Pre(); Out.Write(v); }
	public override void Write(int v) { Pre(); Out.Write(v); }
	public override void Write(long v) { Pre(); Out.Write(v); }
	public override void Write(object v) { Pre(); Out.Write(v); }
	public override void Write(float v) { Pre(); Out.Write(v); }
	public override void Write(string v) { Pre(); Out.Write(v); }
	public override void Write(uint v) { Pre(); Out.Write(v); }
	public override void Write(ulong v) { Pre(); Out.Write(v); }
	public override void Write(string v, object o) { Pre(); Out.Write(v, o); }
	public override void Write(string v, params object[] o) { Pre(); Out.Write(v, o); }
	public override void Write(char[] v, int a, int b) { Pre(); Out.Write(v, a, b); }
	public override void Write(string v, object a, object b) { Pre(); Out.Write(v, a, b); }
	public override void Write(string v, object a, object b, object c) { Pre(); Out.Write(v, a, b, c); }

	public override void WriteLine() { Pre(); Out.WriteLine(); nl = true; }
	
	public override void WriteLine(bool v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(char v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(char[] v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(decimal v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(double v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(int v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(long v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(object v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(float v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(string v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(uint v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(ulong v) { Pre(); Out.WriteLine(v); nl = true; }
	public override void WriteLine(string v, object o) { Pre(); Out.WriteLine(v, o); nl = true; }
	public override void WriteLine(string v, params object[] o) { Pre(); Out.WriteLine(v, o); nl = true; }
	public override void WriteLine(char[] v, int a, int b) { Pre(); Out.WriteLine(v, a, b); nl = true; }
	public override void WriteLine(string v, object a, object b) { Pre(); Out.WriteLine(v, a, b); nl = true; }
	public override void WriteLine(string v, object a, object b, object c) { Pre(); Out.WriteLine(v, a, b, c); nl = true; }

	private void Pre()
	{
		if (nl) {
			Out.Write(PrefixString);
			for (uint i = 0; i < this.Level; i++) {
				Out.Write(IndentString);
			}
			nl = false;
		}
	}
}

} // namespace

File: Implementation/Java/Source/text-api/IndentPrinter.java
package cks.io;

import java.io.IOException;
import java.io.OutputStream;
import java.io.Writer;

/**
 * <p>
 * Printer that manages indentation.  Helps when pretty-printing a
 * hierarchy.  For example:
 * </p>
 *
 * <pre>
 * IndentPrinter p = new IndentPrinter(out, "  ");
 * p.println("Root 1");
 * p.indent();
 * p.println("Child 1");
 * p.println("Child 2");
 * p.dedent();
 * p.println("Root 2");
 * </pre>
 *
 * <p>
 * Yields:
 * </p>
 *
 * <pre>
 * Root 1
 *   Child 1
 *   Child 2
 * Root 2
 * </pre>
 *
 * <p>
 * This class never buffers any data, so you don't need to use {@link #flush()}
 * to force anything out.  Calling {@link #flush()} will just flush on the
 * underlying Writer.
 * </p>
 */
public final class IndentPrinter extends Writer implements Appendable
{
	private final Writer out;
	private final String prefix;
	private final String indent;
	private int numIndents;

	private boolean nl;     ///< If we're currently at the beginning of a new line

	/** Two spaces. */
	public static final String DefaultIndent = "  ";

	/**
	 * @param indent
	 *    The string to add to the prefix for each additional indent
	 *    level.
	 *
	 * @param prefix
	 *    The indentation to prefix to the beginning of each line.
	 *
	 * @param freshLine
	 *    If we're starting on a "fresh" line, we need to add indentation
	 *    to the very first thing printed.  If false, then we'll only start
	 *    adding indentation after the first 'println'.
	 */
	public IndentPrinter(Writer out, String indent, String prefix, boolean freshLine)
	{
		this.out = out;
		this.indent = indent;
		this.prefix = prefix;
		this.numIndents = 0;
		this.nl = freshLine;
	}

   /**
	* Equivalent to: <tt>IndentPrinter(out, indent, prefix, true)</tt>.
	*/
	public IndentPrinter(Writer out, String indent, String prefix)
	{
		this(out, indent, prefix, true);
	}

   /**
	* Equivalent to: <tt>IndentPrinter(out, indent, "")</tt>.
	*/
	public IndentPrinter(Writer out, String indent)
	{
		this(out, indent, "");
	}

   /**
	* Equivalent to: <tt>IndentPrinter(out, IndentPrinter.DefaultIndent)</tt>.
	*/
	public IndentPrinter(Writer out)
	{
		this(out, DefaultIndent);
	}

	public IndentPrinter(OutputStream out, String indent, String prefix, boolean freshLine)
	{
		this(StringUtil.bufferedUtf8Writer(out), indent, prefix, freshLine);
	}

	public IndentPrinter(OutputStream out, String indent, String prefix)
	{
		this(out, indent, prefix, true);
	}

	public IndentPrinter(OutputStream out, String indent)
	{
		this(out, indent, "");
	}

	public IndentPrinter(OutputStream out)
	{
		this(out, DefaultIndent);
	}

	// -----------------------------------------
	// indentation control

	/**
	 * Increase the indentation level by 'numIndents' indentations.
	 */
	public void indent(int numIndents)
	{
		this.numIndents += numIndents;
	}

	/**
	 * Increase the indentation level by one.
	 */
	public void indent()
	{
		this.numIndents += 1;
	}

	/**
	 * Decrease the indentation level by 'numIndents' indentations.
	 */
	public void dedent(int numIndents)
	{
		if (numIndents > this.numIndents) throw new IllegalArgumentException("attempt to decrease indent by " + numIndents + ", but indent level is currently " + this.numIndents);
		this.numIndents -= numIndents;
	}

	/**
	 * Decrease the indentation level by one.
	 */
	public void dedent()
	{
		if (this.numIndents < 1) throw new IllegalArgumentException("attempt to decrease indent, but indent level is currently " + this.numIndents);
		this.numIndents--;
	}

	// -----------------------------------------
	// print functions

	public void print(String v)  throws IOException { pre(); out.write(v == null ? "null" : v); }
	public void print(char[] v)  throws IOException { pre(); out.write(v); }
	public void print(Object v)  throws IOException { pre(); out.write(String.valueOf(v)); }
	public void print(char v)    throws IOException { pre(); out.write(v); }
	public void print(int v)     throws IOException { pre(); out.write(String.valueOf(v)); }
	public void print(long v)    throws IOException { pre(); out.write(String.valueOf(v)); }
	public void print(double v)  throws IOException { pre(); out.write(String.valueOf(v)); }
	public void print(float v)   throws IOException { pre(); out.write(String.valueOf(v)); }
	public void print(boolean v) throws IOException { pre(); out.write(String.valueOf(v)); }
	public void print(CharSequence v) throws IOException { pre(); out.write(v == null ? "null" : v.toString()); }

	public IndentPrinter append(CharSequence v) throws IOException
	{
		print(v);
		return this;
	}

	public IndentPrinter append(CharSequence v, int start, int end) throws IOException
	{
		print(v.subSequence(start, end));
		return this;
	}

	public IndentPrinter append(char c)
		throws IOException
	{
		print(c);
		return this;
	}

	// -----------------------------------------
	// print-ln functions

	public void println(String v)  throws IOException { this.print(v); nl(); }
	public void println(char[] v)  throws IOException { this.print(v); nl(); }
	public void println(Object v)  throws IOException { this.print(v); nl(); }
	public void println(char v)    throws IOException { this.print(v); nl(); }
	public void println(int v)     throws IOException { this.print(v); nl(); }
	public void println(long v)    throws IOException { this.print(v); nl(); }
	public void println(double v)  throws IOException { this.print(v); nl(); }
	public void println(float v)   throws IOException { this.print(v); nl(); }
	public void println(boolean v) throws IOException { this.print(v); nl(); }
	public void println(CharSequence v) throws IOException { out.write(v == null ? "null" : v.toString()); nl(); }

	public void println() throws IOException { nl(); }

	// -----------------------------------------
	// raw write functions
	
	public void write(char[] buf)                   throws IOException { pre(); out.write(buf); }
	public void write(char[] buf, int off, int len) throws IOException { pre(); out.write(buf, off, len); }
	public void write(char c)                       throws IOException { pre(); out.write(c); }
	public void write(String s)                     throws IOException { pre(); out.write(s); }
	public void write(String s, int off, int len)   throws IOException { pre(); out.write(s, off, len); }

	// -----------------------------------------
	// indentation functions

	private void pre() throws IOException
	{
		if (nl) {
			out.write(prefix);
			for (int i = 0; i < this.numIndents; i++) {
				out.write(indent);
			}

			nl = false;
		}
	}

	public static final String NL = System.getProperty("line.separator");

	private void nl() throws IOException
	{
		out.write(NL);
		nl = true;
	}

	// -----------------------------------------
	// java.io.Writer functions

	public void flush() throws IOException { out.flush(); }
	public void close() throws IOException { out.close(); }
}
